core/src/main/scala/org/apache/spark/sql/rapids/tool/util/ToolsPlanGraph.scala

```diff
@@ -16,16 +16,410 @@
+    nodeToStageCluster.put(node.id, clusters)
+    orphanNodes -= node
+  }
+
+  /**
+   * Commits a wholeStageNode to a cluster.
+   * A WholeStageNode is visited after its children are. If any of the children is not assigned to
+   * a cluster, the wNode will transfer its assignment to the child.
+   * @param wNode the wholeStageCodeGen node to be visited
+   * @param orphanNodes the list of nodes that are not assigned to any cluster.
+   * @param clusters the clusterId to assign the node to.
+   * @return true if a change is made.
+   */
+  private def commitNodeToStageCluster(
+    wNode: SparkPlanGraphCluster,
+    orphanNodes: mutable.ArrayBuffer[SparkPlanGraphNode],
+    clusters: Set[Int]): Boolean = {
+    if (nodeToStageCluster.contains(wNode.id) && clusters.subsetOf(nodeToStageCluster(wNode.id))) {
+      // Node is assigned to the same cluster before. Nothing to be done.
+      false
+    } else {
+      val newClusterIds =
+        clusters ++ nodeToStageCluster.getOrElse(wNode.id, ToolsPlanGraph.EMPTY_CLUSTERS)
+      // remove the wNode from orphanNodes if it exists
+      removeNodeFromOrphans(wNode, orphanNodes, newClusterIds)
+      // assign the children to the same clusters if any of them is not assigned already.
+      wNode.nodes.foreach { childNode =>
+        // TODO assert that the child node is not assigned something different than the wStage
+        if (!nodeToStageCluster.contains(childNode.id)) {
+          // assign the child node to the same stage of wNode and remove it from orphans
+          removeNodeFromOrphans(childNode, orphanNodes, newClusterIds)
+        }
+      }
+      true
+    }
+  }
+
+  /**
+   * Assign a node to a clusterId. This method is used to assign a node to a clusterId during the
+   * first visit.
+   * @param node sparkNode to be assigned
+   * @param orphanNodes the list of nodes that are not assigned to any cluster
+   * @param clusters the clusterIds to assign the node to
+   * @return true if a change is made.
+   */
+  private def commitNodeToStageCluster(
+    node: SparkPlanGraphNode,
+    orphanNodes: mutable.ArrayBuffer[SparkPlanGraphNode],
+    clusters: Set[Int]): Boolean = {
+    node match {
+      case cluster: SparkPlanGraphCluster =>
+        // codeGens are special because they propagate their assignment to children nodes.
+        commitNodeToStageCluster(cluster, orphanNodes, clusters)
+      case _ =>
+        removeNodeFromOrphans(node, orphanNodes, clusters)
+        true
+    }
+  }
+
+  /**
+   * Get all the wholeStageNodes in the graph.
+   * @return a list of wholeStageNodes if any.
+   */
+  def getWholeStageNodes(): Seq[SparkPlanGraphCluster] = {
+    nodes.collect {
+      case cluster: SparkPlanGraphCluster if cluster.isInstanceOf[SparkPlanGraphCluster] =>
+        cluster
+    }
+  }
+
+  /**
+   * Walk through the graph nodes and assign them to the correct stage cluster.
+   */
+  protected def assignNodesToStageClusters(): Unit = {
```

Comment:
I was thinking if we could split the steps in this functions into separate functions. Since there is flags shared between different steps, I am fine with this.

Reply:
I agree that's a good idea to break it down.
I prefer to do that as a future work once the heuristics become stable and we had more insights that it works fine across the spectrum.